openedx_external_enrollments/external_enrollments/pathstream_external_enrollment.py

```diff
@@ -1,22 +1,180 @@
+            aws_secret_access_key=settings.OOE_PATHSTREAM_S3_SECRET_KEY,
+        )
+
+    def _download_file(self):
+        """Download the file content from an S3 Bucket.
+
+        :return: file's content (bytes)"""
+        try:
+            response = self.client.get_object(
+                Bucket=self.S3_BUCKET,
+                Key=self.S3_FILE,
+            )
+        except ClientError as error:
+            raise PathstreamTaskExecutionError(
+                '_download_file',
+                'Failed to download the content of the file from S3. Reason: {}'.format(str(error))
+            )
+
+        LOG.info('File content [%s] was downloaded from S3 for [%s]', self.S3_FILE, self.__str__())
+
+        return response['Body'].read()
+
+    def _prepare_new_content(self, content):
+        """This method returns the new content ready to be uploaded to S3. It first sorts the input list
+        by date/time and then encode it since this is the way to upload the file content to S3.
+
+        :param: content (list of strings).
+            The elements of content are expected to be the following comma-separated format:
+            "course,email,2021-08-05 17:53:41.492901,false"
+
+        :return: bytes made up of the data passed in content.
+
+        :raise: AttributeError, IndexError or ValueError in case at least one element in content has a different
+        format than expected."""
+        try:
+            content.sort(
+                #  Extract datetime out of every element in content to be the sorting parameter.
+                key=lambda data: datetime.strptime(data.split(',')[2], settings.OOE_PATHSTREAM_S3_DATETIME_FORMAT)
+            )
+        except (AttributeError, IndexError, ValueError) as error:
+            raise PathstreamTaskExecutionError(
+                '_prepare_new_content',
+                'An enrollment data is invalid. Reason: {} \nNew data: {}'.format(str(error), content)
+            )
+
+        # Convert elements of content to bytes.
+        content = map(lambda data: data.encode(), content)
+
+        return b''.join(content)
+
+    def _upload_file(self, file_content=None):
+        """Upload the file to an S3 bucket using MD5 to ensure that data is not corrupted
+        traversing the network.
+
+        :param: file_content (bytes)."""
+        if not (isinstance(file_content, bytes) and file_content):
+            raise PathstreamTaskExecutionError(
+                '_upload_file', 'Can\'t upload a file with empty content or invalid content type.'
+            )
+
+        hash_algorithm_md5 = hashlib.md5(file_content).digest()
+        content_encoded = base64.b64encode(hash_algorithm_md5)
+
+        try:
+            response = self.client.put_object(
+                Bucket=self.S3_BUCKET,
+                Key=self.S3_FILE,
+                Body=file_content,
+                ContentMD5=content_encoded.decode(),  # String - The base64-encoded 128-bit MD5 digest of the message.
```

Comment:
Why do you need to decode the encoded data? @Jacatove

Reply:
`content_encoded` is already the correct value for the contentMD5 parameter, but it has to be passed as an string, so I use decode() to pass the string representation of `content_encoded`. 

base64 is classified as a binary transform: bytes-like object to bytes mappings. Therefore I had to add the decode step. I found the discussion about this topic here: https://stackoverflow.com/questions/42773386/why-does-base64-b64encode-return-a-bytes-object
@Squirrel18

Reply:
The actual reason for that is because content_encoded is in bytes...